### PR TITLE
[RLlib] Add support for MultiCategorical and MultiActionDistribution for RL Modules and introduce Distribution.to_deterministic()

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

After introducing a couple of action distributions with https://github.com/ray-project/ray/pull/32906, this PR fills in  MultiActionDistributions and MultiCategoricals.

Needs https://github.com/ray-project/ray/pull/32906 (already merged)

Central to this PR is a method `Distribution.get_partial_dist_cls()` that creates partial functions of the underlying distribution - specifically constructor and from_logits. This makes it so that we can create these distributions Algorithm runtime with `.from_logits()`, without knowing what the structure of these Multi-  distributions are.

### CODE REVIEW

File: rllib/core/models/catalog.py [hunk 1]
```diff
@@ -13,9 +16,100 @@
...
 )
 from ray.rllib.models import MODEL_DEFAULTS
+from ray.rllib.models.distributions import Distribution
 from ray.rllib.models.utils import get_filter_config
 from ray.rllib.utils.error import UnsupportedSpaceException
 from ray.rllib.utils.spaces.simplex import Simplex
+from ray.rllib.utils.spaces.space_utils import flatten_space
+
+
+def _multi_action_dist_partial_helper(
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/catalog.py:26]: This is a helper method to create the `MultiActionDistribution` partials.
We use this for tests and the `Catalog. get_dist_cls_from_action_space()` method.
If a user wishes to create distributions that exceed this method of the catalog, they should overwrite it entirely to represent their use-case and don't have to rely on this wonky method.

File: rllib/core/models/catalog.py [hunk 2]
```diff
@@ -313,11 +407,11 @@ def get_dist_cls_from_action_space(
         *,
         framework: Optional[str] = None,
         deterministic: Optional[bool] = False,
-    ) -> Mapping[str, Any]:
+    ) -> Distribution:
         """Returns a distribution class for the given action space.
 
         You can get the required input dimension for the distribution by calling
-        `action_dict_cls.required_model_output_shape(action_space, model_config_dict)`
+        `action_dict_cls.required_input_dim(action_space)`
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/catalog.py:414]: @kouroshHakha As you pointed out once, we don't have to have shapes here. Further more, the multi distributions would become quite complex because the whole "input_lens" logic around splitting inputs would not work anymore.

File: rllib/models/tf/tf_distributions.py [hunk 3]
```diff
@@ -279,13 +277,259 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+
+    @override(Distribution)
+    def sample(self) -> TensorType:
+        arr = [cat.sample() for cat in self.cats]
+        sample_ = tf.stack(arr, axis=1)
+        return sample_
+
+    @override(Distribution)
+    def rsample(self, sample_shape=()):
+        # TODO (Kourosh) Implement Categorical sampling using grad-passthrough trick.
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_distributions.py:309]: This TODO is a copy of the todo found above, in the categorical distribution.
I'm not adding the code here because we can't test it without `Categorical.rsample()`.

File: rllib/models/tf/tf_distributions.py [hunk 4]
```diff
@@ -110,8 +108,21 @@ def __init__(
         logits: tf.Tensor = None,
         temperature: float = 1.0,
     ) -> None:
+        # We assert this here because to_deterministic makes this assumption.
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_distributions.py:111]: Before, the torch distribution made this assumption. I'm adding this code because this class makes the same assumption now.

File: rllib/models/tf/tf_distributions.py [hunk 5]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+    def required_input_dim(space: gym.Space, **kwargs) -> int:
+        return int(np.prod(space.shape, dtype=np.int32))
 
     @classmethod
     @override(Distribution)
     def from_logits(cls, logits: TensorType, **kwargs) -> "TfDeterministic":
         return TfDeterministic(loc=logits)
+
+    def to_deterministic(self) -> "TfDeterministic":
+        return self
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_distributions.py:302]: I thought about throwing an error here and am open to change this.

File: rllib/models/torch/torch_distributions.py [hunk 6]
```diff
@@ -264,13 +272,266 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+            TorchCategorical(logits=logits, temperature=temperature)
+            for logits, temperature in zip(
+                torch.split(logits, input_lens, dim=1), temperatures
+            )
+        ]
+
+        return TorchMultiCategorical(categoricals=categoricals)
+
+    def to_deterministic(self) -> "TorchMultiDistribution":
+        return TorchMultiDistribution([cat.to_deterministic() for cat in self.cats])
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/torch/torch_distributions.py:376]: Note that this PR does not test to_deterministic for most distributions. Lmk if I should add such tests.

**Comment by ollie-iterators** [comment]
The book lint error would be merged by rebasing the branch on the master branch.

File: rllib/models/tf/tf_distributions.py [hunk 7]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+        if isinstance(value, (tf.Tensor, np.ndarray)):
+            split_indices = []
+            for dist in self.flat_child_distributions:
+                if isinstance(dist, TfCategorical):
+                    split_indices.append(1)
+                elif (
+                    isinstance(dist, TfMultiCategorical)
+                    and dist.action_space is not None
+                ):
+                    split_indices.append(np.prod(dist.action_space.shape))
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:440]: I don't understand how this line works?

File: rllib/models/tf/tf_distributions.py [hunk 8]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+                else:
+                    sample = dist.sample()
+                    # Cover Box(shape=()) case.
+                    if len(sample.shape) == 1:
+                        split_indices.append(1)
+                    else:
+                        split_indices.append(tf.shape(sample)[1])
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:447]: also don't understand how this works. hardcoding index 1 seems weird.

File: rllib/models/tf/tf_distributions.py [hunk 9]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+            split_value = tf.split(value, split_indices, axis=1)
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:448]: hard coding axis=1 also seems weird. Does this function work for when the input is batched? (B, d) ?

File: rllib/models/tf/tf_distributions.py [hunk 10]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+        # Structured or flattened (by single action component) input.
+        else:
+            split_value = tree.flatten(value)
+
+        def map_(val, dist):
+            # Remove extra dimension if present.
+            shape = val.shape
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:455]: why are you storing val.shape into a variable shape?

File: rllib/models/tf/tf_distributions.py [hunk 11]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+
+            return dist.logp(val)
+
+        # Remove extra categorical dimension and take the logp of each
+        # component.
+        flat_logps = tree.map_structure(
+            map_, split_value, self.flat_child_distributions
+        )
+
+        return functools.reduce(lambda a, b: a + b, flat_logps)
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:467]: Why don't you simply call sum on flat_logps?

Comment (kouroshHakha) [reply rllib/models/tf/tf_distributions.py:467]: how is this different?

```
import functools

# Define a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use functools.reduce to calculate the sum of the numbers
sum_using_reduce = functools.reduce(lambda a, b: a + b, numbers)
print(sum_using_reduce) # Output: 15

# Use sum to calculate the sum of the numbers
sum_using_sum = sum(numbers)
print(sum_using_sum) # Output: 15
```

File: rllib/models/tf/tf_distributions.py [hunk 12]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+
+    @override(Distribution)
+    def kl(self, other):
+        kl_list = [
+            d.kl(o)
+            for d, o in zip(
+                self.flat_child_distributions, other.flat_child_distributions
+            )
+        ]
+        return functools.reduce(lambda a, b: a + b, kl_list)
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:477]: same thing here?

File: rllib/models/tf/tf_distributions.py [hunk 13]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+
+    @override(Distribution)
+    def entropy(self):
+        entropy_list = [d.entropy() for d in self.flat_child_distributions]
+        return functools.reduce(lambda a, b: a + b, entropy_list)
```
Comment (kouroshHakha) [inline rllib/models/tf/tf_distributions.py:482]: same thing

File: rllib/models/torch/torch_distributions.py [hunk 14]
```diff
@@ -5,14 +5,16 @@
 """
 import gymnasium as gym
 import numpy as np
-from typing import Optional
+from typing import Optional, List, Mapping, Iterable, Dict
```
Comment (kouroshHakha) [inline rllib/models/torch/torch_distributions.py:8]: same comments left on the tf version applies to the torch version.

**Review by kouroshHakha — COMMENTED** [review]
forgot these.

File: rllib/models/tf/tf_distributions.py (hunk 4 above)
Comment (sven1977) [reply rllib/models/tf/tf_distributions.py:111]: What if both are None? Then the assert would pass.
Better:
```
assert (probs is None) != (logits is None)
```

File: rllib/models/tf/tf_distributions.py [hunk 15]
```diff
@@ -126,14 +137,12 @@ def _get_tf_distribution(
...
-        space: gym.Space, model_config: ModelConfigDict
-    ) -> Tuple[int, ...]:
-        return (int(space.n),)
+    def required_input_dim(space: gym.Space, **kwargs) -> int:
+        return int(space.n)
 
     @override(TfDistribution)
     def _rsample(self, sample_shape=()):
-        # TODO (Kourosh) Implement Categorical sampling using grrad-passthrough trick.
+        # TODO (Kourosh) Implement Categorical sampling using grad-passthrough trick.
```
Comment (sven1977) [inline rllib/models/tf/tf_distributions.py:145]: Let's put this in here already:
```
probs = self.probs if self.probs is not None else tf.nn.softmax(self.logits)
return tf.stop_gradient(self._dist.sample(sample_shape) - probs) + probs
```

File: rllib/models/tf/tf_distributions.py [hunk 16]
```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
...
+                vectors to be passed into each child distribution.
+            temperatures: A list of floats representing the temperature to use for
+                each Categorical distribution. If not provided, 1.0 is used for all.
+            **kwargs: Forward compatibility kwargs.
+        """
+        if not temperatures:
+            # If temperatures are not provided, use 1.0 for all actions.
+            temperatures = [1.0] * len(input_lens)
+
+        categoricals = [
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_distributions.py:380]: Thanks good catch!

File: rllib/core/models/catalog.py
Comment (kouroshHakha) [inline rllib/core/models/catalog.py:1]: A couple of things I notice in this file that better to be fixed now:
1. The `self.action_dist_class_fn` seems to be an implementation detail, so it should be private. The actual public api is `self.get_action_dist_cls ` which is all readers of this class should care about. 
2. The following error message is using .format() syntax which is inconsistent and verbose compared to f-string that we are trying to standardize across rllib.
```
UnsupportedSpaceException(
                        "Action space has multiple dimensions "
                        "{}. ".format(action_space.shape)
                        + "Consider reshaping this into a single dimension, "
                        "using a custom action distribution, "
                        "using a Tuple action space, or the multi-agent API."
                    )
```
3. The deterministic flag of `get_dist_cls_from_action_space`  seems redundant after we added `to_deterministic()` api inside the distribution classes. All we need to do now is to get the distribution based on the action space and then call `to_deterministic()` on it.

**Review by kouroshHakha — COMMENTED** [review]
Some more comments.

File: rllib/models/torch/torch_distributions.py [hunk 17]
```diff
@@ -105,27 +105,26 @@ def __init__(
...
+
         self.probs = probs
         self.logits = logits
+        self.temperature = temperature
         self.one_hot = torch.distributions.one_hot_categorical.OneHotCategorical(
-            probs=self.probs, logits=self.logits
+            probs=probs
         )
-        super().__init__(probs=probs, logits=logits, temperature=temperature)
+        super().__init__(probs=probs)
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/torch/torch_distributions.py:120]: Had to make this change so that temperature gets applied to logits **before** we construct both distributions, cat and one hot cat.

**Review by kouroshHakha — APPROVED** [review]
Awesome. It's ready to go.

**Comment by gjoliver** [comment]
lint error is not related.
